File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int f(int i, int j, vector<vector<int>>& dp, vector<int>& nums){
        if(i==nums.size()){
            return 0;
        }
        if(dp[i][j+1]!=-1){
            return dp[i][j+1];
        }
        int l=f(i+1, j, dp, nums);
        if(j==-1 || nums[i]>nums[j]){
            l=max(l, 1+f(i+1, i, dp, nums));
        }
        return dp[i][j+1]=l;
    }
    int lengthOfLIS(vector<int>& nums) {
        int n=nums.size();
        vector<vector<int>>dp(n, vector<int>(n+1, -1));
        return f(0, -1, dp, nums);
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n=len(nums)
        dp=[]
        for i in range(n+1):
            dp.append([0]*(n+1)) 
        for i in range(n-1,-1,-1):
            for j in range(i-1,-2,-1):
                l=dp[i+1][j+1];
                if(j==-1 or nums[i]>nums[j]):
                    l=max(l, 1+dp[i+1][i+1])
                
                dp[i][j+1]=l;
            
        
        return dp[0][0];
    
